staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/event_clock.go

```diff
@@ -0,0 +1,220 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fairqueuing
+
+import (
+	"container/heap"
+	"math/rand"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// EventFunc does some work that needs to be done at or after the
+// given time.  All of the work that needs to be done is finished
+// before this function returns and, if there is an associated
+// `sync.WaitGroup` (see NewFakeEventClock), its counter goes to zero;
+// no other work is left to be completed in goroutines.
+type EventFunc func(time.Time)
+
+// SettablePassiveClock allows setting current time of a passive clock
+type SettablePassiveClock interface {
+	clock.PassiveClock
+	SetTime(time.Time)
+}
+
+// EventClock fires event on time
+type EventClock interface {
+	clock.PassiveClock
+	EventAfterDuration(f EventFunc, d time.Duration)
+	EventAfterTime(f EventFunc, t time.Time)
+}
+
+// RealEventClock fires event on real world time
+type RealEventClock struct {
+	clock.RealClock
+}
+
+// EventAfterDuration schedules an EventFunc
+func (RealEventClock) EventAfterDuration(f EventFunc, d time.Duration) {
+	ch := time.After(d)
+	go func() {
+		select {
+		case t := <-ch:
+			f(t)
+		}
+	}()
+}
+
+// EventAfterTime schedules an EventFunc
+func (r RealEventClock) EventAfterTime(f EventFunc, t time.Time) {
+	now := time.Now()
+	d := t.Sub(now)
+	if d <= 0 {
+		go f(now)
+	} else {
+		r.EventAfterDuration(f, d)
+	}
+}
+
+// FakeEventClock is one whose time does not pass implicitly but
+// rather is explicitly set by invocations of its SetTime method
+type FakeEventClock struct {
+	clock.FakePassiveClock
+
+	// waiters is a heap of waiting work, sorted by time
+	waiters     eventWaiterHeap
+	waitersLock sync.RWMutex
+
+	// clientWG may be nil and if not supplies constraints on time
+	// passing in Run.  The Run method will not pick a new time until
+	// this is nil or its counter is zero.
+	clientWG *sync.WaitGroup
+
+	// fuzz is the amount of noise to add to scheduling.  An event
+	// requested to run at time T will run at some time chosen
+	// uniformly at random from the interval [T, T+fuzz]; the upper
+	// bound is exclusive iff fuzz is non-zero.
+	fuzz time.Duration
+
+	// rand is the random number generator to use in fuzzing
+	rand *rand.Rand
+}
+
+type eventWaiterHeap []eventWaiter
+
+var _ heap.Interface = (*eventWaiterHeap)(nil)
+
+type eventWaiter struct {
+	targetTime time.Time
+	f          EventFunc
+}
+
+// NewFakeEventClock constructor.  The given `r *rand.Rand` must
+// henceforth not be used for any other purpose.  If `r` is nil then a
+// fresh one will be constructed, seeded with the current real time.
+// The clientWG can be `nil` and if not is used to let Run know about
+// additional work that has to complete before time can advance.
+func NewFakeEventClock(t time.Time, clientWG *sync.WaitGroup, fuzz time.Duration, r *rand.Rand) *FakeEventClock {
```

Comment:
More comments later, I have to run now; a general request is to split test-only code into separate, obviously denoted packages (or *_test.go files if it won't be needed in multiple places).

Reply:
Yes, I think the FakeEventClock belongs in a separate clock package.  Ultimately it might go into the regular one (https://github.com/kubernetes/apimachinery/tree/master/pkg/util/clock); until then, we should keep it in a more local clock package.

Reply:
I've moved the EventClock and FakeEventClock to
`staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/clock`

Additionally I've moved the other test only code (integrator.go, no-restraint.go, etc.) to 
`staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing`

I'd like to note that the RealEventClock is not used in fairqueuing for now but the idea currently is that the code is in the package future migration.

This change required some additional refactoring as some of the test code (no-restraint.go) depended on the fairqueuing QueueSet and QueueSetFactory types so I additionally split out the Fairqueuing types and the QueueSet logic, adding:
`staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset`

Reply:
I believe this is resolved now.  Marking resolved.